api/python/cellxgene_census/tests/test_open.py

```diff
@@ -92,18 +110,123 @@ def test_open_soma_errors(requests_mock: rm.Mocker) -> None:
         cellxgene_census.open_soma(census_version="does-not-exist")
 
 
+def test_open_soma_uses_correct_mirror(requests_mock: rm.Mocker) -> None:
+    mock_mirrors = {
+        "default": "test-mirror",
+        "test-mirror": {"provider": "S3", "base_uri": "s3://mirror-bucket-1/", "region": "region-1"},
+        "test-mirror-2": {"provider": "S3", "base_uri": "s3://mirror-bucket-2/", "region": "region-2"},
+    }
+    requests_mock.get(CELL_CENSUS_MIRRORS_DIRECTORY_URL, json=mock_mirrors)
+
+    dir = {
+        "latest": "2022-11-01",
+        "2022-11-01": {
+            "release_date": "2022-11-30",
+            "release_build": "2022-11-01",
+            "soma": {
+                "uri": "s3://ignored-bucket/cell-census/2022-11-01/soma/",
+                "relative_uri": "/cell-census/2022-11-01/soma/",
+                "s3_region": "ignored",
+            },
+            "h5ads": {
+                "uri": "s3://ignored-bucket/cell-census/2022-11-01/h5ads/",
+                "relative_uri": "/cell-census/2022-11-01/soma/",
+                "s3_region": "ignored",
+            },
+        },
+    }
+
+    requests_mock.get(CELL_CENSUS_RELEASE_DIRECTORY_URL, json=dir)
+
+    # Verify that the default mirror is used if no mirror is specified
+    with patch("cellxgene_census._open._open_soma") as m:
+        cellxgene_census.open_soma()
+        m.assert_called_once_with(
+            {"uri": "s3://mirror-bucket-1/cell-census/2022-11-01/soma/", "region": "region-1", "provider": "S3"}, None
+        )
+
+    # Verify that the correct mirror is used if a mirror parameter is specified
+    with patch("cellxgene_census._open._open_soma") as m:
+        cellxgene_census.open_soma(mirror="test-mirror-2")
+        m.assert_called_once_with(
+            {"uri": "s3://mirror-bucket-2/cell-census/2022-11-01/soma/", "region": "region-2", "provider": "S3"}, None
+        )
+
+    # Verify that an error is raised if a non existing mirror is specified
+    with patch("cellxgene_census._open._open_soma") as m:
+        with pytest.raises(
+            KeyError,
+            match=re.escape("Mirror not found."),
+        ):
+            cellxgene_census.open_soma(mirror="bogus-mirror")
+
+
+def test_open_soma_rejects_non_s3_mirror(requests_mock: rm.Mocker) -> None:
+    mock_mirrors = {
+        "default": "test-mirror",
+        "test-mirror": {"provider": "GCS", "bucket": "gcs://mirror-bucket-1/"},
```

Comment:
this is outdated code - ie., the mock data structure does not match the actual datastructure ("bucket"?)

Reply:
This wasn't detected due to `provider` being the only field used for the test here. Thanks for the catch!